spec/src/main/asciidoc/property-expression.adoc

```diff
@@ -0,0 +1,76 @@
+= Property Expression
+
+The property expression is in the format of ${a.property}. When resoving the property expression, the value of `a.property` will be used.
+With the property expression, it enables a property refer to another property.
+
+== An example
+
+Consider the following example:
+
+[source,properties]
+----
+server.url=http://${server.host}:${server.port}/hello
+server.host=my.org
+server.port=9080
+----
+
+When looking up `server.url`, the property value will be resoved to `http://my.org:9080/hello`.
+
+== Nested property expression
+
+The property expression can be nested such as the following:
+
+[source,properties]
+----
+server.url=http://${${foo}.${bar}}/hello
+foo=wonder
+bar=how
+wonder.how=nested
+----
+When looking up `server.url`, the property value will be resoved to `http://nested/hello`.
+
+On resolving the nested expression, the inner property expression must be performed first.
+
+== Assign a default value to property expression
+
+Sometimes, the property expression cannot be resolved due to the unexisting config property. 
+Therefore, it is useful to supply a default value for the configuration expression in the format of `${a.property:value}`. The value after `:` will be the default value.
+
+[source,properties]
+----
+server.url=http://${hostname:localhost}:${port:9080}/hello
+----
+
+On resolving `server.url`, since the properties `hostname` and `port` are not defined, the supplied default values will be used. 
+The value of `server.url` will be `http://localhost:9080/hello`.
```

Comment:
I think property expressions (and any other syntax which we define) definitely must be specified using a BNF-style grammar.